### PR TITLE
initial support of CH34x CDC device

### PR DESCRIPTION
**Pull request by IngHK**
this is the initial version to support the CH34x CDC device.
the CH34x is as popular as FTDI or CH34x.
this version supports only RX and TX and the line_coding settings.
modem control handling will follow...

### CODE REVIEW

**Comment by IngHK** [comment]
what is the preferred code style?
image
what tool do you use tool to format the code?

**Comment by hathach** [comment]
> what is the preferred code style? image what tool do you use tool to format the code?

In short it is K&R 2 spaces. Currently it is quite a bit of a mess, I used to like BSD/allman, but make a switch recently. So you wil mixed style here and there. It will take a while to settle. I tend to only change style within file that I work with.

I don't use any auto format tool, but does occasioally auto-format selection line or file using clion with my own custom of c style. Don't worry much about it, I still prefer manual for readig than auto formmat everywhere.

**Comment by IngHK** [comment]
Thanks!
Please wait with this PR!
I'm finishing the CH34x support and I'll get back to you soon.
Due to compatibility issues I had to change large parts and so current commit is no more ready for review.

**Comment by IngHK** [comment]
It would be nice, if you first review my other PRs.
After they are OK and merged, I will adopt them into the CH34x section.

**Comment by hathach** [comment]
> Thanks! Please wait with this PR! I'm finishing the CH34x support and I'll get back to you soon. Due to compatibility issues I had to change large parts and so current commit is no more ready for review.

Which compatibility issue you are encoutering with, I actually want to wrap up this ch340 support since I need it in another project . Initial test giving great result, I have fixed some gcc strict warning regarding siged/unsigned cast and do more with code reviews and testing. If you plan to push more, please pull first, I will probably push a few more commits before merging.

**Comment by IngHK** [comment]
There is no CH340/CH341 datasheet or programming guide available from vendor WCH.
I found 3 sources of linux driver I used as base for my implementation:
1. from vendor WCH (with poor quality)
2. from Linux Kernel
3. FreeBSD

All 3 differ in details and are NOT working with all derivatives on my desk, so I need to build an extract and to do many "try and errors".

**Comment by IngHK** [comment]
I just pushed the current but unfinished version.
I'm currently working on ch34x_set_line_coding().
It's not ready to merge to master, but you can work with and test.
It includes all other open PR from me.

File: src/class/cdc/cdc_host.h [hunk 1]
```diff
@@ -37,7 +37,7 @@
 // Class Driver Configuration
 //--------------------------------------------------------------------+
 
-// Set Line Control state on enumeration/mounted: DTR ( bit 0), RTS (bit 1)
+// Set Line Control state on enumeration/mounted, refer enums CDC_CONTROL_LINE_STATE_RTS, CDC_CONTROL_LINE_STATE_DTR
```
Comment (hathach) [inline src/class/cdc/cdc_host.h:40]: please don't alter these function comment.

File: src/class/cdc/cdc_debug.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+// logging of baudrate
+#define TU_LOG_BAUDRATE(PRE_TEXT,BAUDRATE) \
+  TU_LOG_DRV(PRE_TEXT "Baudrate %lu Bd\r\n", BAUDRATE)
+
+ // logging of control line state
+#define TU_LOG_CONTROL_LINE_STATE(PRE_TEXT,LINE_STATE)            \
+  TU_LOG_DRV(PRE_TEXT "Control Line State RTS%c DTR%c\r\n", \
+    LINE_STATE & CDC_CONTROL_LINE_STATE_RTS ? '+' : '-',           \
+    LINE_STATE & CDC_CONTROL_LINE_STATE_DTR ? '+' : '-' )
```
Comment (hathach) [inline src/class/cdc/cdc_debug.h:59]: please revert these changes.

**Comment by hathach** [comment]
there is some .bak in the latest push since I am in need to get ch340 support, I will do the clean up right away. Please hold until my push, thank you.

**Comment by hathach** [comment]
> I just pushed the current but unfinished version.
> 
> I'm currently working on ch34x_set_line_coding().
> 
> It's not ready to merge to master, but you can work with and test.
> 
> It includes all other open PR from me.

Thank you, I am cherry picking changes and re-added changes that got lost in the latest push. Please wait a bit before making more push. Hopefully we can got it merged asap. I only tested with an usb2uart dongle ch340g, what else are you testing with ?

File: src/class/cdc/cdc_host.c [hunk 3]
```diff
@@ -1171,4 +1261,327 @@
...
+static bool ch34x_write_reg ( cdch_interface_t* p_cdc, uint16_t reg, uint16_t value, tuh_xfer_cb_t complete_cb, uintptr_t user_data )
+{
+  return ch34x_control_out ( p_cdc, CH34X_REQ_WRITE_REG, /* value */ reg, /* index */ value, complete_cb, user_data );
+}
+
+//static bool ch34x_read_reg_request ( cdch_interface_t* p_cdc, uint16_t reg,
+//                                     uint8_t *buffer, uint16_t buffersize, tuh_xfer_cb_t complete_cb, uintptr_t user_data )
+//{
+//  return ch34x_control_in ( p_cdc, CH34X_REQ_READ_REG, reg, /* index */ 0, buffer, buffersize, complete_cb, user_data );
+//}
```
[the remark this answers is not in the document]
Comment (IngHK) [reply src/class/cdc/cdc_host.c:1359]: ok

File: src/class/cdc/cdc_host.h (hunk 1 above)
Comment (IngHK) [reply src/class/cdc/cdc_host.h:40]: ok

File: src/class/cdc/cdc_debug.h (hunk 2 above)
Comment (IngHK) [reply src/class/cdc/cdc_debug.h:59]: it has useful debug logs...

**Comment by IngHK** [comment]
> there is some .bak in the latest push since I am in need to get ch340 support, I will do the clean up right away. Please hold until my push, thank you.
> 
> PS: when merging, please also retain my changes, it seems like your latest push discard most of my changes. But that is OK, there is an conflict going on since both of us are updating the driver.

yes, it was a mistake

**Comment by IngHK** [comment]
> there is some .bak in the latest push since I am in need to get ch340 support, I will do the clean up right away. Please hold until my push, thank you.
> 
> PS: when merging, please also retain my changes, it seems like your latest push discard most of my changes. But that is OK, there is an conflict going on since both of us are updating the driver.

sorry

**Comment by IngHK** [comment]
> > I just pushed the current but unfinished version.
> > I'm currently working on ch34x_set_line_coding().
> > It's not ready to merge to master, but you can work with and test.
> > It includes all other open PR from me.
> 
> Thank you, I am cherry picking changes and re-added changes that got lost in the latest push. Please wait a bit before making more push. Hopefully we can got it merged asap. I only tested with an usb2uart dongle ch340g, what else are you testing with ?

OK.
I tested it with several CH340 and CH341 devices: USB/TTL, USB/RS232 and measurement instrument with integrated CH340.
I tested it with CH340C, CH341A and some other devices, where I'm not able to see the exact letter behind CH340.
But all of them with version 0x30 or newer.

**Comment by hathach** [comment]
> > there is some .bak in the latest push since I am in need to get ch340 support, I will do the clean up right away. Please hold until my push, thank you.
> > PS: when merging, please also retain my changes, it seems like your latest push discard most of my changes. But that is OK, there is an conflict going on since both of us are updating the driver.
> 
> yes, it was a mistake

no problem at all, I did a cherry picking and push the update to the PR. Please pull first before making more update. We got everything update now. 

> OK. I tested it with several CH340 and CH341 devices: USB/TTL, USB/RS232 and measurement instrument with integrated CH340. I tested it with CH340C, CH341A and some other devices, where I'm not able to see the exact letter behind CH340. But all of them with version 0x30 or newer.

Thank you for your information. There is no worry to get it all working in the first PR. We can always fix it later on.

File: src/class/cdc/cdc_debug.h (hunk 2 above)
Comment (hathach) [reply src/class/cdc/cdc_debug.h]: these can be output in simpler form using existing log in right location.There is really no need to have dedicated log for these control signals especially adding new header for this. I removed most of those in latest push. If needed, we can add some to cdc host driver while doing more test.

**Comment by IngHK** [comment]
I reviewed your cherry pick and everything is OK.
I just found some small things in my last commit.

File: src/class/cdc/cdc_host.c [hunk 4]
```diff
@@ -1171,4 +1170,351 @@ static void cp210x_process_config(tuh_xfer_t* xfer) {
...
+            line_coding.parity == CDC_LINE_CODING_PARITY_SPACE) {
+          lcr |= CH34X_LCR_MARK_SPACE;
+        }
+      }
+      TU_ASSERT (line_coding.stop_bits == CDC_LINE_CODING_STOP_BITS_1 ||
+                 line_coding.stop_bits == CDC_LINE_CODING_STOP_BITS_2,); // 1.5 stop bits not supported
+      if (line_coding.stop_bits == CDC_LINE_CODING_STOP_BITS_2) {
+        lcr |= CH34X_LCR_STOP_BITS_2;
+      }
+      TU_ASSERT (ch34x_control_out(p_cdc, CH34X_REQ_SERIAL_INIT, (uint16_t) (lcr << 8 | 0x9c), (uint16_t) (factor << 8 | 0x80 | divisor),
```
Comment (hathach) [inline src/class/cdc/cdc_host.c:1484]: both https://github.com/OLIMEX/ch340-dkms/blob/master/ch340.c#L282 and linux run this commad with (0,0) then set baudrate and lsr afterwards. Should we follow their init sequence as well ?

Comment (IngHK) [reply src/class/cdc/cdc_host.c:1479]: My list of drivers examined above was not complete.
As already mentioned, I examined the Linux sources of:
- Vendor WCH (with poor quality)
- Linux Kernel
- FreeBSD

None of the 3 Linux drivers worked with all of my CH34x on the desk, not even the WCH one!
Then I tried the two different WCH Windows driver
- V3.5.2019.1
- V3.8.2023.2

Only V3.5.2019.1 worked with all my CH34x.
I analyzed both versions with an USB bus analyzer, get the working init and config sequence and overtaken them in my code.
To be clear, the original WCH manufacturer driver for Linux does not work and behaves differently than the original WCH manufacturer driver for Windows.
I may have a clone and not an original WCH CH34x on my desk and WCH may have changed its driver in the meantime so that it no longer works with the clones.
Quite ugly...

To your question:
The fully functioning WCH Windows driver uses the CH34X_REQ_SERIAL_INIT (CMD_C1 in WCH driver) request to configure line coding and baud rate (factor and divisor) at the same time with one control request.
The WCH Linux driver uses the sequence CH34X_REQ_SERIAL_INIT (0,0) and then writing the registers for line coding and baudrate, but as I said, this is also possible together in a CH34X_REQ_SERIAL_INIT.
CH34X_REQ_SERIAL_INIT may perform additional initializations...

After this initial configuration, the driver sends register write requests to the line coding register LCR (0x18) and factor/divisor registers (0x12 and 0x13) if corresponding settings are to be changed during operation.
Some CH34x devices before version 0x30 may also use LCR2 register (0x25), but I've no device on my desk to verify it.
See https://github.com/torvalds/linux/blob/052d534373b7ed33712a63d5e17b2b6cdbce84fd/drivers/usb/serial/ch341.c#L272

Two registers can be written simultaneously as part of a register write request.
wValue contains the addresses of the two registers in low and high bytes and wIndex contains the register values to be written also in low and high bytes.
See https://github.com/freebsd/freebsd-src/blob/a079c891c01b1f8b32d79b490d9e7b7542e26d0f/sys/dev/usb/serial/uchcom.c#L446

**Comment by IngHK** [comment]
I'm now implementing ch34x_set_line_coding() and then and then I'm done with CH34x.
My goal is to add complete support for all popular CDC devices.
And so support for the Prolific PL2303 is missing.
And the line coding settings for FTDI and CP210x are still missing.
What are your plans regarding finishing FTDI and CP210x support?

**Comment by IngHK** [comment]
And are there plans to implement functions to get the line states CTR, DSR, DCD and RI?
=> Abstract Control Model Notifications of CDC PSTN Subclass

File: src/class/cdc/cdc_host.c (hunk 4 above)
Comment (hathach) [reply src/class/cdc/cdc_host.c:1479]: thank you for clarifcation, I also take the peek at linux and wch linux driver. Didn't know there is an windows source driver that init with these 2 extra parameters.

**Comment by IngHK** [comment]
Take care when testing!
You need to unplug (power down) and replug the CH34x!
It has no reset to default via USB control requests...

**Comment by IngHK** [comment]
ch34x_set_line_coding() is finished

File: src/class/cdc/cdc_host.c [hunk 5]
```diff
@@ -1171,4 +1171,383 @@ static void cp210x_process_config(tuh_xfer_t* xfer) {
...
+}
+
+static bool ch34x_set_line_coding(cdch_interface_t* p_cdc, cdc_line_coding_t const* line_coding,
+                                  tuh_xfer_cb_t complete_cb, uintptr_t user_data) {
+  p_cdc->baudrate_requested = line_coding->bit_rate;
+  p_cdc->user_control_cb = complete_cb;
+  uint8_t factor, divisor, lcr;
+  TU_ASSERT (ch34x_get_factor_divisor(line_coding->bit_rate, &factor, &divisor));
+  TU_ASSERT (ch34x_get_lcr(line_coding, &lcr));
+  TU_ASSERT (ch34x_control_out(p_cdc, CH34X_REQ_SERIAL_INIT, (uint16_t) (lcr << 8 | 0x9c), (uint16_t) (factor << 8 | 0x80 | divisor),
```
Comment (hathach) [inline src/class/cdc/cdc_host.c:1475]: thank you for the update, however, I think we should not use init command to change baudrate/link data. Those are just convenient when init. But it will reset every internal state including all control signal e.g dtr/rts etc, while we only want to change e.g baudrate while connectinng. I am still revisig the driver, lackig official documentation make thing more difficult, but I prefer to program lcr and div_prescale as linux approach.

Comment (hathach) [reply src/class/cdc/cdc_host.c:1475]: PS: I am makinngn some morme changes and test regarding set line coding, will push when it is available.

Comment (IngHK) [reply src/class/cdc/cdc_host.c:1475]: I have had the same thoughts and concerns.
WCH just names the request “CMD_C1”.
In this respect, I wonder where the term “serial init” comes from.
This is probably just an interpretation of the non-WCH developers who didn't have an official datasheet.

But no matter, then we have to implement a sequence of 2 register write requests.
I have no experience with this (especially with special usage of user_data) and in the meantime I'm already working on another project.
Can you please provide a code suggestion?
When I have free capacity again in a few days, I will be happy to carry out the final tests with all of my CH34x devices.

Comment (hathach) [reply src/class/cdc/cdc_host.c:1475]: I am adding code for that, will make a push soon enough.

**Comment by hathach** [comment]
I have pushed the update that implement 2 stage set line coding, also added single set_data_format() as new API. Please test it out to see if that works for you. I will probably do a bit more clean up before merging this.

> I'm now implementing ch34x_set_line_coding() and then and then I'm done with CH34x. My goal is to add complete support for all popular CDC devices. And so support for the Prolific PL2303 is missing. And the line coding settings for FTDI and CP210x are still missing. What are your plans regarding finishing FTDI and CP210x support?

> And are there plans to implement functions to get the line states CTS, DSR, DCD and RI? => Abstract Control Model Notifications of CDC PSTN Subclass

missinng features will be implemented when needed and/or having free time.

File: src/class/cdc/cdc_host.c [hunk 6]
```diff
@@ -1410,82 +1414,83 @@ static bool ch34x_get_factor_divisor(uint32_t baval, uint8_t* factor, uint8_t* d
...
       break;
   }
-  *factor = a;
-  *divisor = b;
 
-  return true;
+  // reg divisor = a, reg prescaler = b
+  // According to linux code we need to set bit 7 of UCHCOM_REG_BPS_PRE,
+  // otherwise the chip will buffer data.
+  return (uint16_t) (a << 8 | 0x80 | b);
```
Comment (IngHK) [inline src/class/cdc/cdc_host.c:1623]: "a" is uint8_t
it need to be casted (uint16_t) before shift

File: src/class/cdc/cdc_host.c [hunk 7]
```diff
@@ -1410,82 +1414,83 @@ static bool ch34x_get_factor_divisor(uint32_t baval, uint8_t* factor, uint8_t* d
...
-      *lcr |= CH34X_LCR_CS5;
-      break;
-    case 6:
-      *lcr |= CH34X_LCR_CS6;
-      break;
-    case 7:
-      *lcr |= CH34X_LCR_CS7;
+static inline uint8_t ch34x_get_lcr(uint8_t stop_bits, uint8_t parity, uint8_t data_bits) {
+  uint8_t lcr = CH34X_LCR_ENABLE_RX | CH34X_LCR_ENABLE_TX;
+  TU_VERIFY(data_bits >= 5, 0);
```
Comment (IngHK) [inline src/class/cdc/cdc_host.c:1629]: incomplete check
should be  TU_VERIFY(data_bits >= 5 && data_bits <= 8, 0);

File: src/class/cdc/cdc_host.c [hunk 8]
```diff
@@ -1410,82 +1414,83 @@ static bool ch34x_get_factor_divisor(uint32_t baval, uint8_t* factor, uint8_t* d
...
+  lcr |= (uint8_t) (data_bits - 5);
+
+  if (parity) {
```
Comment (IngHK) [inline src/class/cdc/cdc_host.c:1632]: Better to understand:
```
  switch(parity) {
    case CDC_LINE_CODING_PARITY_NONE:
      break;

    case CDC_LINE_CODING_PARITY_ODD:
      lcr |= CH34X_LCR_ENABLE_PAR;
      break;

    case CDC_LINE_CODING_PARITY_EVEN:
      lcr |= CH34X_LCR_ENABLE_PAR | CH34X_LCR_PAR_EVEN;
      break;

    case CDC_LINE_CODING_PARITY_MARK:
      lcr |= CH34X_LCR_ENABLE_PAR | CH34X_LCR_MARK_SPACE;
      break;

    case CDC_LINE_CODING_PARITY_SPACE:
      lcr |= CH34X_LCR_ENABLE_PAR | CH34X_LCR_MARK_SPACE | CH34X_LCR_PAR_EVEN;
      break;

    default: break;
  }
```

File: src/class/cdc/cdc_host.c [hunk 9]
```diff
@@ -1410,82 +1414,83 @@ static bool ch34x_get_factor_divisor(uint32_t baval, uint8_t* factor, uint8_t* d
       }
       a = (unsigned char) (c / baval);
```
Comment (IngHK) [inline src/class/cdc/cdc_host.c:1609]: `a = (uint8_t) (c / baval);`

**Review by IngHK — COMMENTED** [review]
I advise not to change that.
There are explicit instructions that default line coding is necessary to start up the CH34x.
For this reason I created `CFG_TUH_CDC_LINE_CODING_ON_ENUM_CH34X`.
It may happen, that the following steps 'CONFIG_CH34X_SPECIAL_REG_WRITE' etc. will not work properly.
see
- https://github.com/torvalds/linux/blob/c25b24fa72c734f8cd6c31a13548013263b26286/drivers/usb/serial/ch341.c#L386
- inital 9600 Bd write https://github.com/WCHSoftGroup/ch341ser_linux/blob/8966d44d8c8f7e5178a53c34984f2e74d655ccb3/driver/ch341.c#L218

**Comment by IngHK** [comment]
> I advise not to change that. There are explicit instructions that default line coding is necessary to start up the CH34x. For this reason I created `CFG_TUH_CDC_LINE_CODING_ON_ENUM_CH34X`. It may happen, that the following steps 'CONFIG_CH34X_SPECIAL_REG_WRITE' etc. will not work properly. see
> 
>     * https://github.com/torvalds/linux/blob/c25b24fa72c734f8cd6c31a13548013263b26286/drivers/usb/serial/ch341.c#L386
> 
>     * inital 9600 Bd write https://github.com/WCHSoftGroup/ch341ser_linux/blob/8966d44d8c8f7e5178a53c34984f2e74d655ccb3/driver/ch341.c#L218

When I tried overtaking all three driver sources (WCH, Linux & FreeBSD) one after the other, I found that all 3 did NOT work with all CH34x devices on my desk.
Only the WCH Windows driver V3.5.2019.1 works properly.
I then recorded the USB data traffic of the Windows driver.
The recorded sequences have little in common with those of the 3 Linux sources.
Here are the control sequences:
image

File: src/class/cdc/cdc_host.c [hunk 10]
```diff
@@ -425,11 +422,10 @@ static void cdch_internal_control_complete(tuh_xfer_t* xfer) {
       case SERIAL_DRIVER_FTDI:
         switch (xfer->setup->bRequest) {
           case FTDI_SIO_MODEM_CTRL:
-            p_cdc->line_state = (uint8_t) (tu_le16toh(xfer->setup->wValue) & 0x00ff);
```
[the remark this answers is not in the document]
Comment (hathach) [reply src/class/cdc/cdc_host.c:400]: may or may not, I remembered it cause narrow down warnings from gcc which is treated as error build.

File: src/class/cdc/cdc_host.c [hunk 11]
```diff
@@ -1526,21 +1414,25 @@ static void ch34x_process_config(tuh_xfer_t* xfer) {
...
+
+      // Init CH34x with line coding
       TU_ASSERT (ch34x_control_out(p_cdc, CH34X_REQ_SERIAL_INIT, tu_u16(lcr, 0x9c), div_ps,
                                    ch34x_process_config, CONFIG_CH34X_SPECIAL_REG_WRITE),);
       break;
     }
 
     case CONFIG_CH34X_SPECIAL_REG_WRITE:
       // do special reg write, purpose unknown, overtaken from WCH driver
+      p_cdc->line_coding = line_coding;
```
[the remark this answers is not in the document]
Comment (hathach) [reply src/class/cdc/cdc_host.c]: here we already set the line coding, it should be set right away, since there could be error in the middle steps.

**Comment by hathach** [comment]
> > I advise not to change that. There are explicit instructions that default line coding is necessary to start up the CH34x. For this reason I created `CFG_TUH_CDC_LINE_CODING_ON_ENUM_CH34X`. It may happen, that the following steps 'CONFIG_CH34X_SPECIAL_REG_WRITE' etc. will not work properly. see
> > ```
> > * https://github.com/torvalds/linux/blob/c25b24fa72c734f8cd6c31a13548013263b26286/drivers/usb/serial/ch341.c#L386
> > 
> > * inital 9600 Bd write https://github.com/WCHSoftGroup/ch341ser_linux/blob/8966d44d8c8f7e5178a53c34984f2e74d655ccb3/driver/ch341.c#L218
> > ```
> 
> When I tried overtaking all three driver sources (WCH, Linux & FreeBSD) one after the other, I found that all 3 did NOT work with all CH34x devices on my desk. Only the WCH Windows driver V3.5.2019.1 works properly. I then recorded the USB data traffic of the Windows driver. The recorded sequences have little in common with those of the 3 Linux sources. Here are the control sequences: image

Thank you for your reviews and fixinng up them, everything makes sense except for the line coding on enum. Have you tried testing with ch34x board without set line coding while enum ? I have tested and it worked just fine with application manually set it in mounted callback 

https://github.com/IngHK/tinyusb/blob/cdc_ch34x_support/examples/host/cdc_msc_hid/src/cdc_app.c#L98

For dynamic application such as Arduino, it can have different baudrate for each example (or device it connects to) and a fixed coding does not satify that. For Arduino, it almost always try to set line codinng in the callback, forcinng setting line coding on enum will actualy set line coding twice (enum, then app), which is not neccessary.

Line coding must be set before data transferring, if `CFG_TUH_CDC_LINE_CODING_ON_ENUM` is not defined, then it is application/user's resposibility.

File: src/class/cdc/cdc_host.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (IngHK) [reply src/class/cdc/cdc_host.c]: OK

**Comment by IngHK** [comment]
> > > I advise not to change that. There are explicit instructions that default line coding is necessary to start up the CH34x. For this reason I created `CFG_TUH_CDC_LINE_CODING_ON_ENUM_CH34X`. It may happen, that the following steps 'CONFIG_CH34X_SPECIAL_REG_WRITE' etc. will not work properly. see
> > > ```
> > > * https://github.com/torvalds/linux/blob/c25b24fa72c734f8cd6c31a13548013263b26286/drivers/usb/serial/ch341.c#L386
> > > 
> > > * inital 9600 Bd write https://github.com/WCHSoftGroup/ch341ser_linux/blob/8966d44d8c8f7e5178a53c34984f2e74d655ccb3/driver/ch341.c#L218
> > > ```
> > 
> > 
> > When I tried overtaking all three driver sources (WCH, Linux & FreeBSD) one after the other, I found that all 3 did NOT work with all CH34x devices on my desk. Only the WCH Windows driver V3.5.2019.1 works properly. I then recorded the USB data traffic of the Windows driver. The recorded sequences have little in common with those of the 3 Linux sources. Here are the control sequences: image
> 
> Thank you for your reviews and fixinng up them, everything makes sense except for the line coding on enum. Have you tried testing with ch34x board without set line coding while enum ? I have tested and it worked just fine with application manually set it in mounted callback
> 
> https://github.com/IngHK/tinyusb/blob/cdc_ch34x_support/examples/host/cdc_msc_hid/src/cdc_app.c#L98
> 
> For dynamic application such as Arduino, it can have different baudrate for each example (or device it connects to) and a fixed coding does not satify that. For Arduino, it almost always try to set line codinng in the callback, forcinng setting line coding on enum will actualy set line coding twice (enum, then app), which is not neccessary.
> 
> Line coding must be set before data transferring, if `CFG_TUH_CDC_LINE_CODING_ON_ENUM` is not defined, then it is application/user's resposibility.

OK, I understand your reasons.
I also don't think it's a good thing that the CH34x is already made ready for reception during the enumeration with 9600 8N1.
However, we don't know whether all CH34x will start correctly if no 9600 8N1 is set during the start-up sequence.
Then, if CFG_TUH_CDC_LINE_CODING_ON_ENUM is not defined, the startup sequence should not be carried out during enumeration, but once at the first call to ch34x_set_line_coding(), just like the Windows driver does.

**Comment by hathach** [comment]
> > > > I advise not to change that. There are explicit instructions that default line coding is necessary to start up the CH34x. For this reason I created `CFG_TUH_CDC_LINE_CODING_ON_ENUM_CH34X`. It may happen, that the following steps 'CONFIG_CH34X_SPECIAL_REG_WRITE' etc. will not work properly. see
> > > > ```
> > > > * https://github.com/torvalds/linux/blob/c25b24fa72c734f8cd6c31a13548013263b26286/drivers/usb/serial/ch341.c#L386
> > > > 
> > > > * inital 9600 Bd write https://github.com/WCHSoftGroup/ch341ser_linux/blob/8966d44d8c8f7e5178a53c34984f2e74d655ccb3/driver/ch341.c#L218
> > > > ```
> > > 
> > > 
> > > When I tried overtaking all three driver sources (WCH, Linux & FreeBSD) one after the other, I found that all 3 did NOT work with all CH34x devices on my desk. Only the WCH Windows driver V3.5.2019.1 works properly. I then recorded the USB data traffic of the Windows driver. The recorded sequences have little in common with those of the 3 Linux sources. Here are the control sequences: image
> > 
> > 
> > Thank you for your reviews and fixinng up them, everything makes sense except for the line coding on enum. Have you tried testing with ch34x board without set line coding while enum ? I have tested and it worked just fine with application manually set it in mounted callback
> > https://github.com/IngHK/tinyusb/blob/cdc_ch34x_support/examples/host/cdc_msc_hid/src/cdc_app.c#L98
> > For dynamic application such as Arduino, it can have different baudrate for each example (or device it connects to) and a fixed coding does not satify that. For Arduino, it almost always try to set line codinng in the callback, forcinng setting line coding on enum will actualy set line coding twice (enum, then app), which is not neccessary.
> > Line coding must be set before data transferring, if `CFG_TUH_CDC_LINE_CODING_ON_ENUM` is not defined, then it is application/user's resposibility.
> 
> OK, I understand your reasons. I also don't think it's a good thing that the CH34x is already made ready for reception during the enumeration with 9600 8N1. However, we don't know whether all CH34x will start correctly if no 9600 8N1 is set during the start-up sequence. Then, if `CFG_TUH_CDC_LINE_CODING_ON_ENUM` is not defined, the startup sequence should not be carried out during enumeration, but once at the first call of `ch34x_set_line_coding()`, just like the Windows driver does.

Ok, I agree, setting default for ch34x when CFG_TUH_CDC_LINE_CODING_ON_ENUM not defined is totally OK since it does not take any extra request since it is part of SERIAL_INIT request. I kind of mixed up with linux driver. All is good.

**Comment by IngHK** [comment]
> perfect, the PR is well tested and is optimized for ch34x support. Thank you very much for your patient and following up to my request as well as my modification. I am glad that we are finally getting this merged.

Thank you, I enjoyed working with you and learned a lot.
I will be doing compatibility tests with the master version in the next few days.

**Comment by hathach** [comment]
thank you, I enjoyed co-working with you on this PR as well.
